src/remote/idcert.rs

```diff
@@ -0,0 +1,694 @@
+                Tag::CTX_0, |c| c.skip_u8_if(2)
+            )?;
+
+            let serial_number = Serial::take_from(cons)?;
+            let _sig = SignatureAlgorithm::x509_take_from(cons)?;
+            let issuer = Name::take_from(cons)?;
+            let validity = Validity::take_from(cons)?;
+            let subject = Name::take_from(cons)?;
+            let subject_public_key_info = PublicKey::take_from(cons)?;
+            
+            let extensions = cons.take_constructed_if(
+                Tag::CTX_3, IdExtensions::take_from
+            )?;
+
+            Ok(TbsIdCert {
+                serial_number,
+                issuer,
+                validity,
+                subject,
+                subject_public_key_info,
+                extensions,
+            })
+        })
+    }
+
+    /// Returns an encoder for the value.
+    pub fn encode_ref(&self) -> impl encode::Values + '_ {
+        encode::sequence((
+            (
+                Constructed::new(
+                    Tag::CTX_0,
+                    2.encode(), // Version 3 is encoded as 2
+                ),
+                self.serial_number.encode(),
+                SignatureAlgorithm::default().x509_encode(),
+                self.issuer.encode_ref(),
+            ),
+            (
+                self.validity.encode(),
+                self.subject.encode_ref(),
+                self.subject_public_key_info.clone().encode(),
+                self.extensions.encode_ref(),
+            ),
+        ))
+    }
+}
+
+/// # Creation and Conversion
+///
+impl TbsIdCert {
+    
+    /// Creates an TbsIdCert to be signed with the Signer trait.
+    /// 
+    /// Note that this function is private - it is used by the specific
+    /// public functions for creating a TA ID cert, or CMS EE ID cert.
+    fn new(
+        serial_number: Serial,
+        validity: Validity,
+        issuing_key: &PublicKey,
+        subject_key: &PublicKey,
+        extensions: IdExtensions,
+    ) -> TbsIdCert {
+        let issuer = Name::from_pub_key(issuing_key);
+        let subject = Name::from_pub_key(subject_key);
+
+        TbsIdCert {
+            serial_number,
+            issuer,
+            validity,
+            subject,
+            subject_public_key_info: subject_key.clone(),
+            extensions,
+        }
+    }
+
+    /// Converts the value into a signed ID certificate.
+    fn into_cert<S: Signer>(
+        self,
+        signer: &S,
+        key: &S::KeyId,
+    ) -> Result<IdCert, SigningError<S::Error>> {
+        let data = Captured::from_values(Mode::Der, self.encode_ref());
+        let signature = signer.sign(key, SignatureAlgorithm::default(), &data)?;
+        Ok(IdCert {
+            signed_data: SignedData::new(data, signature),
+            tbs: self
+        })
+    }
+
+
+}
+
+//------------ IdExtensions --------------------------------------------------
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct IdExtensions {
```

Comment:
Is there a reason to have this as a separate type rather than just as part of `TbsIdCert`?

Reply:
done, can you please check? There is a lot more overlap now with the code for `Cert` but because of subtleties is probably still easier to live with duplication at this point.